synapse/config/_util.py

```diff
@@ -70,9 +71,11 @@ def json_error_to_config_error(
 Model = TypeVar("Model", bound=BaseModel)
 
 
+# Here is where I need help with the Type. I put in TypeAlias because I know it works,
+# for what I suspect is the wrong reason.
 def parse_and_validate_mapping(
     config: Any,
-    model_type: Type[Model],
+    model_type: TypeAlias,
```

Comment:
TODO

Reply:
If this is reverted back to `Type[Model]`, the following lints are thrown:

```
synapse/config/_util.py:91: error: Variable "model_type" is not valid as a type  [valid-type]
synapse/config/_util.py:91: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
synapse/config/workers.py:284: error: Argument 2 to "parse_and_validate_mapping" has incompatible type "<typing special form>"; expected "Union[Type[InstanceTCPLocationConfig], Type[InstanceUNIXLocationConfig]]"  [arg-type]
```

Reply:
Which is why I need help with the Typing. I was hoping @DMRobertson would be kind enough to lend a hand with that, as he had before.

Reply:
@realtyem It seems like a couple of us have at least had a cursory try at addressing this with no success. Probably best to add `# type: ignore[arg-type]` with our best explanation on what might be going on. And if we're lucky cross-referencing a relevant mypy bug: https://github.com/python/mypy/issues

Reply:
For full disclosure(and to help with how extensive my confusion is), here is the full set of mypy readouts if you remove all the `# type: ignore` around this.
```log
synapse/config/_util.py:89: error: Variable "model_type" is not valid as a type  [valid-type]
synapse/config/_util.py:89: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
synapse/config/workers.py:284: error: Argument 2 to "parse_and_validate_mapping" has incompatible type "<typing special form>"; expected "Union[Type[InstanceTcpLocationConfig], Type[InstanceUnixLocationConfig]]"  [arg-type]
```

I was hoping this would be a more useful generic function that could process other 'maps' as they are migrated to Pydantic. The website mentioned [in the error message](https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases) was where I got the `TypeAlias` from. I'm not sure it's a bug, really.